.github/workflows/differential-shellcheck.yml

```diff
@@ -0,0 +1,28 @@
+---
+
+name: Differential ShellCheck
+on:
+  pull_request:
+    branches: [ master ]
+
+permissions:
+  contents: read
+
+jobs:
+  lint:
+    runs-on: ubuntu-22.04
+
+    permissions:
+      security-events: write
+      pull-requests: write
+
+    steps:
+      - name: Repository checkout
+        uses: actions/checkout@2541b1294d2704b0964813337f33b291d3f8596b
```

Comment:
Why do you use a specific commit hash?

Reply:
Since it is good security practice, based on [Security hardening for GitHub Actions](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions).

Downside being that you would need to add also something like Dependabot.

Reply:
I am a noob with using bots that act as if they were regular users.
In particular I had never let computers act on their own on
as if they were me, cf.
https://github.com/rear/rear/issues/2645#issuecomment-878090407
```
Via the tokens
I give the programmers of the SCM
...
permissions
to do things in my name
because what actually happens is not what I myself do
but what programmers of those workflows/operations implement
where I have no control in practice what actually happens
(in practice I cannot verify their source code).
```

If my understanding is right then it seems
"something like Dependabot to update SHA
 when there is a new release of an action"
contradicts the idea of
"pinning an action to a full length commit SHA"
because such a "Dependabot" would blindly change
that SHA which also contradicts in
https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions
```
* Audit the source code of the action

Ensure that the action is handling the content
of your repository and secrets as expected.
For example, check that secrets are not sent
to unintended hosts, or are not inadvertently logged.
```
So when the source code of the action was audited
only its specific SHA is trused and this SHA must
not be changed unless a new version was also audited.

Reply:
> If my understanding is right then it seems "something like Dependabot to update SHA when there is a new release of an action" contradicts the idea of "pinning an action to a full length commit SHA" because such a "Dependabot" would blindly change that SHA which also contradicts in https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions

Dependabot will only create PR that will change SHA, it's upon maintainers, if they will inspect new commits/source code. If you found some insufficiencies in the new version, you could close Dependabot PR or leave it open and Dependabot will let you know if there is a new version.

Example of Dependabot PR: https://github.com/systemd/systemd/pull/24168

I generally take Dependabot as a good way to get notification when a new version of my dependency is released.

> So when the source code of the action was audited only its specific SHA is trused and this SHA must not be changed unless a new version was also audited.

From a security point of view, yes. On GitHub there are repositories, that uses only pinning to commit SHA (e.g. [systemd](https://github.com/systemd/systemd) and [dracut](https://github.com/dracutdevs/dracut)) and there are repositories that uses standard "potentially unsafe" pining to tags (e.g. [probot](https://github.com/probot/probot), [super-linter](https://github.com/github/super-linter) and [pi-hole](https://github.com/pi-hole/pi-hole)). So it's something to consider. I personally use pinning to tags in all of my projects, since it's more readable, but less secure.

Reply:
@jamacku 
thank your for your explanation how Dependabot works